sdk/inc/azure/core/az_json.h

```diff
@@ -58,141 +58,88 @@ typedef struct
  */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_number(double value)
-{
-  return (az_json_token){
-    .kind = AZ_JSON_TOKEN_NUMBER,
-    ._internal.number = value,
-  };
-}
+AZ_NODISCARD az_result
+az_json_token_get_uint64(az_json_token const* json_token, uint64_t* out_value);
 
 /**
- * @brief az_json_token_string returns an az_json_token containing a string.
+ * @brief Returns the JSON token's number as a 32-bit unsigned integer.
  *
- * @param value A span over a string indicating how the az_json_token should be initialized.
+ * @param json_token A pointer to an #az_json_token instance.
+ * @param out_value A pointer to a variable to receive the value.
+ * @return AZ_OK if the number is returned.<br>
+ * AZ_ERROR_JSON_INVALID_STATE if the kind != AZ_JSON_TOKEN_NUMBER.
  */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_string(az_span value)
-{
-  return (az_json_token){
-    .kind = AZ_JSON_TOKEN_STRING,
-    ._internal.string = value,
-  };
-}
+AZ_NODISCARD az_result
+az_json_token_get_uint32(az_json_token const* json_token, uint32_t* out_value);
 
 /**
- * @brief az_json_token_string returns an az_json_token containing an object.
+ * @brief Returns the JSON token's string after unescaping it, if required.
  *
- * @param value A span over an object indicating how the az_json_token should be initialized.
- */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_object(az_span value)
-{
-  return (az_json_token){
-    .kind = AZ_JSON_TOKEN_BEGIN_OBJECT,
-    ._internal.span = value,
-  };
-}
-
-/**
- * @brief returns an az_json_token representing the start of an object.
- */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_begin_object()
-{
-  return (az_json_token){ .kind = AZ_JSON_TOKEN_BEGIN_OBJECT, ._internal = { 0 } };
-}
-
-/**
- * @brief returns an az_json_token representing the end of an object.
- */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_end_object()
-{
-  return (az_json_token){ .kind = AZ_JSON_TOKEN_END_OBJECT, ._internal = { 0 } };
-}
-
-/**
- * @brief returns an az_json_token representing the start of an array.
+ * @param json_token A pointer to an #az_json_token instance.
+ * @param destination A pointer to a buffer where the string should be copied into.
+ * @param destination_max_size The maximum available space within the buffer referred to by
+ * \p destination.
+ * @param[out] out_string_length __[nullable]__ Contains the number of bytes written to the \p
+ * destination which denote the length of the unescaped string. If `NULL` is passed, the parameter
+ * is ignored.
+ * @return AZ_OK if the string is returned.<br>
+ * AZ_ERROR_JSON_INVALID_STATE if the kind != AZ_JSON_TOKEN_STRING.<br>
+ * AZ_ERROR_INSUFFICIENT_SPAN_SIZE if \p destination does not have enough size.
  */
-AZ_NODISCARD AZ_INLINE az_json_token az_json_token_begin_array()
-{
-  return (az_json_token){ .kind = AZ_JSON_TOKEN_BEGIN_ARRAY, ._internal = { 0 } };
-}
+AZ_NODISCARD az_result az_json_token_get_string(
```

Comment:
@JeffreyRichter - I know we discussed adding such an API on span directly, rather than az_json_token, but 
having the `az_json_token_get_boolean` API felt odd to add on span. Also, all the call sites ended up needing to fetch the slice field from the token every time, which impacts usability.

Most calling code doesn't first check the token kind before calling these APIs, and want to fail for mismatched/unexpected token kind anyway.

Let me know what you think.

Reply:
Also, I decided against adding an AZ_JSON_TOKEN_ESCAPED_STRING token kind since we would also need to add a AZ_JSON_TOKEN_ESCAPED_PROPERTY_NAME token kind and that exposes too much implementation detail complexity to the caller.